util/logger/src/metric.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright 2018 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use std::thread;
+use std::time::Instant;
+
+use parking_lot::RwLock;
+use time;
+
+pub type MetricKey = String;
+
+pub struct MetricEntry {
+    pub value: u64,
+    pub prev: u64,
+}
+
+impl MetricEntry {
+    fn new_with(initial: u64) -> Self {
+        MetricEntry {
+            value: initial,
+            prev: 0,
+        }
+    }
+
+    fn changed(&self) -> u64 {
+        self.value - self.prev
+    }
+
+    fn inc(&mut self) {
+        self.value += 1;
+    }
+
+    fn reset_prev(&mut self) {
+        self.prev = self.value;
+    }
+}
+
+pub struct Metric {
+    inner: Arc<RwLock<MetricInner>>,
```

Comment:
Make `Metric` have `RwLock<BtreeMap<MetricKey, AtomicUsize>>`.

Reply:
Since `start_thread(&self)` function needs to clone `inner` to use it in the other thread, I added the `Arc`. Could you know any other method to use `inner` or `self` in other thread?

Reply:
What I commented was making `increase` function hold read lock to make multiple threads can increase the count.
So I think `Arc<RwLock<...>>` is okay.

Reply:
Aha, I understand it. It would be great.